File: app/ui/dialogs/export_dialog.py
import os
from PySide6.QtCore import Qt
from PySide6.QtWidgets import (QDialog, QVBoxLayout, QHBoxLayout, QLabel,
                             QComboBox, QPushButton, QSlider, QLineEdit, QFileDialog, QSpinBox, QCheckBox, QGroupBox)
from app.core.image_document import ImageDocument
from app.workers.export_worker import ExportWorker
from app.utils.settings import settings
from app.utils.logger import logger

class ExportDialog(QDialog):
    """Dialog window for configuring image export options."""

    def __init__(self, document: ImageDocument, parent=None):
        super().__init__(parent)
        self.doc = document
        self.setWindowTitle("Export Image")
        self.setFixedSize(480, 420)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(16, 16, 16, 16)
        layout.setSpacing(14)

        # Output File Location
        grp_file = QGroupBox("Export Destination", self)
        vbox_file = QVBoxLayout(grp_file)
        hbox_path = QHBoxLayout()
        self.txt_path = QLineEdit(self)
        default_dir = settings.get("export_folder") or os.path.expanduser("~")
        default_file = os.path.join(default_dir, "export_nobg.png")
        self.txt_path.setText(default_file)
        hbox_path.addWidget(self.txt_path)

        btn_browse = QPushButton("Browse...", self)
        btn_browse.clicked.connect(self._browse_destination)
        hbox_path.addWidget(btn_browse)
        vbox_file.addLayout(hbox_path)
        layout.addWidget(grp_file)

        # Export Format & Quality
        grp_fmt = QGroupBox("Format & Quality", self)
        vbox_fmt = QVBoxLayout(grp_fmt)

        hbox_fmt = QHBoxLayout()
        hbox_fmt.addWidget(QLabel("Format:", self))
        self.combo_format = QComboBox(self)
        self.combo_format.addItems(["PNG", "JPG", "WEBP"])
        self.combo_format.currentTextChanged.connect(self._on_format_changed)
        hbox_fmt.addWidget(self.combo_format)
        vbox_fmt.addLayout(hbox_fmt)

        # Quality slider (for JPG/WEBP)
        self.hbox_qual = QHBoxLayout()
        self.hbox_qual.addWidget(QLabel("Quality:", self))
        self.lbl_qual_val = QLabel("95%", self)
        self.hbox_qual.addWidget(self.lbl_qual_val)
        self.slider_quality = QSlider(Qt.Orientation.Horizontal, self)
        self.slider_quality.setRange(1, 100)
        self.slider_quality.setValue(95)
        self.slider_quality.valueChanged.connect(lambda v: self.lbl_qual_val.setText(f"{v}%"))
        self.hbox_qual.addWidget(self.slider_quality)
        vbox_fmt.addLayout(self.hbox_qual)
        layout.addWidget(grp_fmt)

        # Dimension Resizing
        grp_dim = QGroupBox("Resolution / Dimensions & Presets", self)
        vbox_dim = QVBoxLayout(grp_dim)

        hbox_preset = QHBoxLayout()
        hbox_preset.addWidget(QLabel("Preset:", self))
        self.combo_preset = QComboBox(self)
        self.combo_preset.addItems([
            "Current Canvas Size",
            "YouTube Thumbnail (1920×1080)",
            "YouTube Shorts / Reels (1080×1920)",
            "Instagram Post (1080×1080)",
            "Instagram Story (1080×1920)",
            "Custom"
        ])
        self.combo_preset.currentIndexChanged.connect(self._on_preset_changed)
        hbox_preset.addWidget(self.combo_preset, stretch=1)
        vbox_dim.addLayout(hbox_preset)

        hbox_spins = QHBoxLayout()
        orig_w = self.doc.width() if self.doc else 1920
        orig_h = self.doc.height() if self.doc else 1080

        hbox_spins.addWidget(QLabel("Width:", self))
        self.spin_w = QSpinBox(self)
        self.spin_w.setRange(1, 10000)
        self.spin_w.setValue(orig_w)
        hbox_spins.addWidget(self.spin_w)

        hbox_spins.addWidget(QLabel("Height:", self))
        self.spin_h = QSpinBox(self)
        self.spin_h.setRange(1, 10000)
        self.spin_h.setValue(orig_h)
        hbox_spins.addWidget(self.spin_h)
        vbox_dim.addLayout(hbox_spins)

        self.chk_aspect = QCheckBox("Lock Aspect Ratio", self)
        self.chk_aspect.setChecked(True)
        vbox_dim.addWidget(self.chk_aspect)
        layout.addWidget(grp_dim)


        # Action Buttons
        hbox_btns = QHBoxLayout()
        hbox_btns.addStretch()
        btn_cancel = QPushButton("Cancel", self)
        btn_cancel.clicked.connect(self.reject)
        hbox_btns.addWidget(btn_cancel)

        self.btn_export = QPushButton("💾 Export Now", self)
        self.btn_export.setObjectName("btn_primary")
        self.btn_export.clicked.connect(self._do_export)
        hbox_btns.addWidget(self.btn_export)
        layout.addLayout(hbox_btns)

        self._on_format_changed("PNG")

    def _on_preset_changed(self, idx: int):
        presets = {
            0: (self.doc.width() if self.doc else 1920, self.doc.height() if self.doc else 1080),
            1: (1920, 1080),
            2: (1080, 1920),
            3: (1080, 1080),
            4: (1080, 1920)
        }
        if idx in presets:
            w, h = presets[idx]
            self.spin_w.setValue(w)
            self.spin_h.setValue(h)

    def _on_format_changed(self, fmt: str):

        is_lossy = fmt in ("JPG", "WEBP")
        self.slider_quality.setEnabled(is_lossy)

        # Update extension
        path = self.txt_path.text()
        base = os.path.splitext(path)[0]
        ext = f".{fmt.lower()}"
        self.txt_path.setText(base + ext)

    def _browse_destination(self):
        fmt = self.combo_format.currentText()
        filter_str = f"{fmt} Images (*.{fmt.lower()})"
        file_path, _ = QFileDialog.getSaveFileName(self, "Save Image", self.txt_path.text(), filter_str)
        if file_path:
            self.txt_path.setText(file_path)

    def _do_export(self):
        path = self.txt_path.text()
        fmt = self.combo_format.currentText()
        qual = self.slider_quality.value()
        w = self.spin_w.value()
        h = self.spin_h.value()

        self.btn_export.setEnabled(False)
        self.worker = ExportWorker(self.doc, path, format_str=fmt, quality=qual, width=w, height=h)
        self.worker.finished.connect(self._on_finished)
        self.worker.error.connect(self._on_error)
        self.worker.start()

    def _on_finished(self, out_path: str):
        self.accept()

    def _on_error(self, err_msg: str):
        self.btn_export.setEnabled(True)
        print(f"Export Error: {err_msg}")

File: src/ui/dialogs/ExportDialog.cpp
#include "ui/dialogs/ExportDialog.h"
#include "workers/ExportWorker.h"
#include "utils/Settings.h"
#include "utils/Logger.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QGroupBox>
#include <QFileDialog>
#include <QDir>
#include <QMessageBox>

namespace ImageCut {
namespace UI {

ExportDialog::ExportDialog(std::shared_ptr<Core::ImageDocument> doc, QWidget* parent)
    : QDialog(parent), m_doc(doc)
{
    setWindowTitle("Export Image");
    setFixedSize(480, 420);

    QVBoxLayout* layout = new QVBoxLayout(this);
    layout->setContentsMargins(16, 16, 16, 16);
    layout->setSpacing(14);

    QGroupBox* grpFile = new QGroupBox("Export Destination", this);
    QVBoxLayout* vboxFile = new QVBoxLayout(grpFile);
    QHBoxLayout* hboxPath = new QHBoxLayout();
    m_txtPath = new QLineEdit(this);
    QString defaultDir = Utils::Settings::getInstance().get("export_folder", QDir::homePath());
    m_txtPath->setText(defaultDir + "/export_nobg.png");
    hboxPath->addWidget(m_txtPath);

    QPushButton* btnBrowse = new QPushButton("Browse...", this);
    connect(btnBrowse, &QPushButton::clicked, [this]() {
        QString fmt = m_comboFormat->currentText();
        QString filter = QString("%1 Images (*.%2)").arg(fmt).arg(fmt.toLower());
        QString path = QFileDialog::getSaveFileName(this, "Save Image", m_txtPath->text(), filter);
        if (!path.isEmpty()) m_txtPath->setText(path);
    });
    hboxPath->addWidget(btnBrowse);
    vboxFile->addLayout(hboxPath);
    layout->addWidget(grpFile);

    QGroupBox* grpFmt = new QGroupBox("Format & Quality", this);
    QVBoxLayout* vboxFmt = new QVBoxLayout(grpFmt);

    QHBoxLayout* hboxFmt = new QHBoxLayout();
    hboxFmt->addWidget(new QLabel("Format:", this));
    m_comboFormat = new QComboBox(this);
    m_comboFormat->addItems({ "PNG", "JPG", "WEBP" });
    connect(m_comboFormat, &QComboBox::currentTextChanged, [this](const QString& fmt) {
        bool isLossy = (fmt == "JPG" || fmt == "WEBP");
        m_sliderQuality->setEnabled(isLossy);
    });
    hboxFmt->addWidget(m_comboFormat);
    vboxFmt->addLayout(hboxFmt);

    QHBoxLayout* hboxQual = new QHBoxLayout();
    hboxQual->addWidget(new QLabel("Quality:", this));
    m_lblQualVal = new QLabel("95%", this);
    hboxQual->addWidget(m_lblQualVal);
    m_sliderQuality = new QSlider(Qt::Horizontal, this);
    m_sliderQuality->setRange(1, 100);
    m_sliderQuality->setValue(95);
    connect(m_sliderQuality, &QSlider::valueChanged, [this](int v) {
        m_lblQualVal->setText(QString("%1%").arg(v));
    });
    hboxQual->addWidget(m_sliderQuality);
    vboxFmt->addLayout(hboxQual);
    layout->addWidget(grpFmt);

    QGroupBox* grpDim = new QGroupBox("Resolution / Dimensions & Presets", this);
    QVBoxLayout* vboxDim = new QVBoxLayout(grpDim);

    QHBoxLayout* hboxPreset = new QHBoxLayout();
    hboxPreset->addWidget(new QLabel("Preset:", this));
    m_comboPreset = new QComboBox(this);
    m_comboPreset->addItems({
        "Current Canvas Size",
        "YouTube Thumbnail (1920×1080)",
        "YouTube Shorts / Reels (1080×1920)",
        "Instagram Post (1080×1080)",
        "Instagram Story (1080×1920)",
        "Custom"
    });
    connect(m_comboPreset, QOverload<int>::of(&QComboBox::currentIndexChanged), [this](int idx) {
        int origW = m_doc ? m_doc->canvasWidth : 1920;
        int origH = m_doc ? m_doc->canvasHeight : 1080;
        if (idx == 0) { m_spinW->setValue(origW); m_spinH->setValue(origH); }
        else if (idx == 1) { m_spinW->setValue(1920); m_spinH->setValue(1080); }
        else if (idx == 2) { m_spinW->setValue(1080); m_spinH->setValue(1920); }
        else if (idx == 3) { m_spinW->setValue(1080); m_spinH->setValue(1080); }
        else if (idx == 4) { m_spinW->setValue(1080); m_spinH->setValue(1920); }
    });
    hboxPreset->addWidget(m_comboPreset, 1);
    vboxDim->addLayout(hboxPreset);

    QHBoxLayout* hboxSpins = new QHBoxLayout();
    int origW = m_doc ? m_doc->canvasWidth : 1920;
    int origH = m_doc ? m_doc->canvasHeight : 1080;

    hboxSpins->addWidget(new QLabel("Width:", this));
    m_spinW = new QSpinBox(this);
    m_spinW->setRange(1, 10000);
    m_spinW->setValue(origW);
    hboxSpins->addWidget(m_spinW);

    hboxSpins->addWidget(new QLabel("Height:", this));
    m_spinH = new QSpinBox(this);
    m_spinH->setRange(1, 10000);
    m_spinH->setValue(origH);
    hboxSpins->addWidget(m_spinH);
    vboxDim->addLayout(hboxSpins);

    m_chkAspect = new QCheckBox("Lock Aspect Ratio", this);
    m_chkAspect->setChecked(true);
    vboxDim->addWidget(m_chkAspect);
    layout->addWidget(grpDim);

    QHBoxLayout* hboxBtns = new QHBoxLayout();
    hboxBtns->addStretch();
    QPushButton* btnCancel = new QPushButton("Cancel", this);
    connect(btnCancel, &QPushButton::clicked, this, &QDialog::reject);
    hboxBtns->addWidget(btnCancel);

    m_btnExport = new QPushButton("💾 Export Now", this);
    m_btnExport->setObjectName("btn_primary");
    connect(m_btnExport, &QPushButton::clicked, this, &ExportDialog::doExport);
    hboxBtns->addWidget(m_btnExport);
    layout->addLayout(hboxBtns);
}

void ExportDialog::doExport() {
    if (!m_doc) return;
    QString path = m_txtPath->text();
    QString fmt = m_comboFormat->currentText();
    int qual = m_sliderQuality->value();
    int w = m_spinW->value();
    int h = m_spinH->value();

    m_btnExport->setEnabled(false);
    auto worker = new Workers::ExportWorker(m_doc, path, fmt, qual, w, h);
    
    // Safely delete worker object ONLY after QThread has completely stopped running!
    connect(worker, &QThread::finished, worker, &QObject::deleteLater);

    connect(worker, &Workers::ExportWorker::exportFinished, this, [this](const QString& outPath) {
        LOG_INFO("Export completed: " + outPath.toStdString());
        accept();
    }, Qt::QueuedConnection);

    connect(worker, &Workers::ExportWorker::error, this, [this](const QString& err) {
        LOG_ERROR("Export Error: " + err.toStdString());
        m_btnExport->setEnabled(true);
        QMessageBox::critical(this, "Export Failed", "Could not export image:\n" + err);
    }, Qt::QueuedConnection);

    worker->start();
}

} // namespace UI
} // namespace ImageCut
